sklearn/utils/_testing.py

```diff
@@ -671,6 +689,162 @@ def check_docstring_parameters(func, doc=None, ignore=None):
+    objects : list of {classes, functions, data descriptors, NumpyDocString}
+        Objects to check.
+        May be `NumpyDocString` instances or objects (classes, functions or
+        data descriptors) with docstrings that can be parsed by numpydoc.
+
+    include_params : list of str or bool, default=False
+        List of parameters to be included. If True, all parameters are included,
+        if False, checking is skipped for parameters.
+        Can only be set if `exclude_params` is None.
+
+    exclude_params : list of str or None, default=None
+        List of parameters to be excluded. If None, no parameters are excluded.
+        Can only be set if `include_params` is True.
+
+    include_attribs : list of str or bool, default=False
+        List of attributes to be included. If True, all attributes are included,
+        if False, checking is skipped for attributes.
+        Can only be set if `exclude_attribs` is None.
+
+    exclude_attribs : list of str or None, default=None
+        List of attributes to be excluded. If None, no attributes are excluded.
+        Can only be set if `include_attribs` is True.
+
+    include_returns : list of str or bool, default=False
+        List of returns to be included. If True, all returns are included,
+        if False, checking is skipped for returns.
+        Can only be set if `exclude_returns` is None.
+
+    exclude_returns : list of str or None, default=None
+        List of returns to be excluded. If None, no returns are excluded.
+        Can only be set if `include_returns` is True.
+
+    Examples
+    --------
+    >>> from sklearn.metrics import (mean_absolute_error, mean_squared_error,
+    ... median_absolute_error)
+    >>> from sklearn.utils.testing import assert_docstring_consistency
+    ... # doctest: +SKIP
+    >>> assert_docstring_consistency([mean_absolute_error, mean_squared_error],
+    ... include_params=['y_true', 'y_pred', 'sample_weight'])  # doctest: +SKIP
+    >>> assert_docstring_consistency([median_absolute_error, mean_squared_error],
+    ... include_params=True)  # doctest: +SKIP
+    """
+    from numpydoc import docscrape
+
+    Args = namedtuple("args", ["include", "exclude", "arg_name"])
+    section_dict = {
+        "Parameters": Args(include_params, exclude_params, "params"),
+        "Attributes": Args(include_attribs, exclude_attribs, "attribs"),
+        "Returns": Args(include_returns, exclude_returns, "returns"),
+    }
+    skip_sections = []
+    for section, args in section_dict.items():
+        if args.exclude and args.include is not True:
+            raise TypeError(
+                f"The 'exclude_{args.arg_name}' argument can be set only when the "
+                f"'include_{args.arg_name}' argument is True."
+            )
+        if args.include is False:
+            skip_sections.append(section)
+
+    # Skip section by removing from dictionary
+    for section in skip_sections:
+        del section_dict[section]
```

Comment:
I don't love having to loop again here. I could alternatively loop through a copy of `section_dict`:

```python
for section, args in section_dict.copy().items():
```
and then del items from `section_dict`, but is that bad practice?

Reply:
Perhaps you can alternatively do

```python
for section, args in list(section_dict.items()):
```

which iterates over a copy instead of the iterator itself?

Reply:
My understanding is that, using `dict.copy` and `list(dict.items())` are effectively both just making shallow copies if the values of the dict are mutables: they do not perform deepcopy, but instead when the item is removed from the original dict the actual mutable is still there because reference count has not dropped to zero. Shouldn't be that expensive either way.

```python
>>> import gc  # garbage collector module
>>> d = {"a": [1, 2, 3], "b": [4, 5, 6]}
>>> it = list(d.items())
>>> it
[('a', [1, 2, 3]), ('b', [4, 5, 6])]
>>> id(it[0][1])
2060607087936
>>> id(d["a"])  # same id, so no copy is made
2060607087936
>>> gc.get_referrers(it[0][1])  # two referrers: the dict and the list of items
[{'a': [1, 2, 3], 'b': [4, 5, 6]}, ('a', [1, 2, 3])]
>>> del d["a"]
>>> gc.get_referrers(it[0][1])  # one referrer: only the list of items
[('a', [1, 2, 3])]
>>> it[0][1]  # can still be accessed
[1, 2, 3]
```

Reply:
Interesting, thanks!